src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,316 @@
+                    // If the symbol's range is in the parent symbol's range
+                    if (symbol.Range.Start.Line > currentRange.Start.Line && symbol.Range.End.Line < currentRange.End.Line)
+                    {
+                        currentGroup.Add(symbol);
+                    }
+                    else
+                    {
+                        // Push existing group
+                        documentSymbolGroups.Add(currentGroup.ToImmutableAndFree());
+                        // Create new group with this symbol as the parent
+                        currentGroup = ArrayBuilder<DocumentSymbol>.GetInstance(1, symbol);
+                        currentRange = symbol.Range;
+                    }
+                }
+
+                documentSymbolGroups.Add(currentGroup.ToImmutableAndFree());
+                return documentSymbolGroups.ToImmutableAndFree();
+            }
+
+            // Given a flat array containing a Document Symbol and its descendants, returns the Document Symbol
+            // with its descendants recursively nested. The first Document Symbol in the array is considered the root node.
+            static DocumentSymbol CreateDocumentSymbolTree(ImmutableArray<DocumentSymbol> documentSymbols)
+            {
+                var node = documentSymbols.First();
+                var childDocumentSymbols = documentSymbols.RemoveAt(0);
+                node.Children = GroupDocumentSymbolTrees(childDocumentSymbols)
+                    .Select(group => CreateDocumentSymbolTree(group))
+                    .ToArray();
+                return node;
+            }
+        }
+
+        /// <summary>
+        /// Converts an array of type DocumentSymbol to an array of type DocumentSymbolItem.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolItem> GetDocumentSymbolItems(DocumentSymbol[] documentSymbols)
+        {
+            var documentSymbolItems = ArrayBuilder<DocumentSymbolItem>.GetInstance();
+            if (documentSymbols.Length == 0)
+            {
+                return documentSymbolItems.ToImmutableAndFree();
+            }
+
+            foreach (var documentSymbol in documentSymbols)
+            {
+                var documentSymbolItem = new DocumentSymbolItem(documentSymbol);
+                if (documentSymbol.Children is not null)
+                    documentSymbolItem.Children = GetDocumentSymbolItems(documentSymbol.Children);
+                documentSymbolItems.Add(documentSymbolItem);
+            }
+
+            return documentSymbolItems.ToImmutableAndFree();
+        }
+
+        public static async Task<JToken?> DocumentSymbolsRequestAsync(
+            ITextBuffer textBuffer,
+            ILanguageServiceBrokerShim languageServiceBroker,
+            string? textViewFilePath,
+            CancellationToken cancellationToken)
+        {
+            if (textViewFilePath is null)
+                return null;
```

Comment:
nit: i would prefer we not even get to this point if file-path is null.  caller can/should check.